### PR TITLE
Eliminate second copy of `Relation` type

### PR DESCRIPTION
**Pull request by mitchellwrosen**
## Overview

This PR eliminates `U.Util.Relation` in favor of `Unison.Util.Relation`. The primary motivation, discussed briefly in #2455, was to avoid having to port API/performance improvements to both places.

## Implementation notes

It was a bit of a roundabout process but I ultimately discovered removing `U.Util.Relation` would have been trivial - it's not even _used_ anywhere :)

But, as part of moving towards unifying the two types, I made the changes in this PR, which I think are a slight improvement to the package structure, so I didn't necessarily want to throw them away before discussing.

Concretely:

- `Unison.Prelude` was moved into a `lib/unison-prelude` package
- `Unison.Util.Relation`, `Unison.Util.Relation3`, and `Unison.Util.Relation4` were moved into a `lib/unison-util-relation` package
- The `Unison.Hashable` instance for `Relation{3,4}` was moved into `Unison.Hashable`, so the `Relation` types can now be considered "hackagey", i.e. not containing anything Unison-specific.

Another implementation note: I formatted a couple modules along the way

### CODE REVIEW

**Comment by aryairani** [comment]
> it's not even _used_ anywhere :)

Oops 😅 

@mitchellwrosen WDYT about the changes to `hie.yaml`; I figure that's hie the output of `gen-hie` but why does it add `./` to every path, and why does it list each individual test file? I have been cleaning it up manually each time.

**Review by aryairani — APPROVED** [review]
I left a note about hie.yaml but otherwise lgtm.

**Comment by mitchellwrosen** [comment]
Eeewwww... yeah, why did it do that?! The trailing `./` I could live with, but explicitly listing source files... yeesh.

I'll clean it manually :sweat_smile:

File: lib/unison-util-relation/src/Unison/Util/Relation4.hs [hunk 1]
```diff
@@ -116,7 +114,3 @@ instance (Ord a, Ord b, Ord c, Ord d) => Monoid (Relation4 a b c d) where
     d2' = Map.unionWith (<>) (d2 s1) (d2 s2)
     d3' = Map.unionWith (<>) (d3 s1) (d3 s2)
     d4' = Map.unionWith (<>) (d4 s1) (d4 s2)
-
-instance (H.Hashable d1, H.Hashable d2, H.Hashable d3, H.Hashable d4)
-       => H.Hashable (Relation4 d1 d2 d3 d4) where
-  tokens s = [ H.accumulateToken $ toNestedList s ]
```
Comment (ChrisPenner) [inline lib/unison-util-relation/src/Unison/Util/Relation4.hs:122]: ~Just curious, why was this removed?~

Nvm, I see it was just moved due to dependency changes 👍🏼

**Review by ChrisPenner — APPROVED** [review]
All looks good to me; 
Which formatter are you using btw? We should probably all sync up on that.

We should keep an eye on our package splitting to see which configurations perform best, splitting off packages which rarely change is almost always a good thing, but we should be careful about splitting off too many packages where a change requires editing both, or just too many packages in general.

HLS really doesn't like it when you make changes in a dependant package for the module you're currently editing, and stack will generally force a complete rebuild of all dependents in that case.

That said, I think prelude and relations are generally okay to split off 👍🏼

**Comment by aryairani** [comment]
> Which formatter are you using btw? We should probably all sync up on that.

`ormolu` is what we're moving towards currently.

**Comment by mitchellwrosen** [comment]
> We should keep an eye on our package splitting to see which configurations perform best, splitting off packages which rarely change is almost always a good thing, but we should be careful about splitting off too many packages where a change requires editing both, or just too many packages in general.

Totally. In my experience, generally speaking, fewer packages are preferred for the spaghetti guts of an application. But anything that could conceivably exist on Hackage I think can be nice to pull out to the top level, because then its implementation can actually be hidden behind a (hopefully) well-documented interface!
